File: Parallelism/python/parallelism.py
import requests
import json 
from multiprocessing import Process
import random
import time
from multiprocessing import Pool
import tracemalloc

def http_request(request_count):
    id_num = random.randint(1,100)
    res = requests.get('https://dummyjson.com/todos/' + str(id_num))
    response = json.loads(res.text)
    # print(response)

            
if __name__ == '__main__':
    tracemalloc.start()  # start tracing memory allocations
    start_time = time.perf_counter()
    with Pool() as pool:
        result = pool.map(http_request, range(0,25))
    finish_time = time.perf_counter()
    print("Program finished in {} seconds - using multiprocessing".format(finish_time-start_time))
    print("---")

    # start_time = time.perf_counter()
    # for i in range(100):
    #     http_request(100)
    # finish_time = time.perf_counter()
    # print("Program finished in {} seconds".format(finish_time-start_time))
    # print("---")

    current, peak = tracemalloc.get_traced_memory()
    print(f"Peak memory usage: {peak / 1024 / 1024:.2f} MB")
    tracemalloc.stop()  # stop tracing memory allocations

File: Parallelism/go/parallelism.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"time"
	//"sync"
	"runtime"
)

func httpRequest() {
	idNum := rand.Intn(100) + 1

	resp, err := http.Get(fmt.Sprintf("https://dummyjson.com/todos/%d", idNum))
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	var result map[string]interface{}
	err = json.Unmarshal(body, &result)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Uncomment the following line to print the response
	 fmt.Println(result)
}

func main() {
	var m runtime.MemStats

	runtime.ReadMemStats(&m) // read current memory statistics
	numRequests := 75

	// Parallel execution using goroutines and a WaitGroup
	// startTime := time.Now()
	// var wg sync.WaitGroup

	// for i := 0; i < numRequests; i++ {
	// 	wg.Add(1)
	// 	go func() {
	// 		defer wg.Done()
	// 		httpRequest()
	// 	}()
	// }

	// wg.Wait()
	// elapsedTime := time.Since(startTime)
	// fmt.Printf("Total execution time: %.6f seconds\n", elapsedTime.Seconds())
	// fmt.Println("---")

	// Sequential execution
	startTime := time.Now()
	for i := 0; i < numRequests; i++ {
		httpRequest()
	}
	elapsedTime := time.Since(startTime)
	fmt.Printf("Total execution time: %.6f seconds\n", elapsedTime.Seconds())
	fmt.Println("---")

	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)

	peakMemoryUsage := float64(memStats.HeapSys) / (1024 * 1024)
	fmt.Printf("Peak memory usage: %.2f MB\n", peakMemoryUsage)
}
